### DISCUSSION TITLE
Cannot define multiple columns within the same table that have the same relationship to another table

### DISCUSSION THREAD

**Discussion by graysonhead** [SeaQL/sea-orm#251] [General]
Hey there, currently migrating one of my projects over from Diesel to take advantage of tokio and I'm running into a problem with one of my tables.

Here is the table in question:

(Irrelevant columns omitted for brevity)

```sql
CREATE TABLE attackers (
    attacker_id BIGINT unsigned PRIMARY KEY UNIQUE NOT NULL AUTO_INCREMENT,
    ship_type_id BIGINT unsigned,
    weapon_type_id BIGINT unsigned,
    FOREIGN KEY (ship_type_id) REFERENCES esi_types (type_id),
    FOREIGN KEY (weapon_type_id) REFERENCES esi_types (type_id)
);
```

The generated `attackers.rs` entity is generated as so:

```rust
#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    CharacterPublicInfo,
    #[sea_orm(
        belongs_to = "super::esi_types::Entity",
        from = "Column::ShipTypeId",
        to = "super::esi_types::Column::TypeId",
        on_update = "Restrict",
        on_delete = "Restrict"
    )]
    EsiTypes,
    #[sea_orm(
        belongs_to = "super::esi_types::Entity",
        from = "Column::WeaponTypeId",
        to = "super::esi_types::Column::TypeId",
        on_update = "Restrict",
        on_delete = "Restrict"
    )]
    EsiTypes,
}
```

This looks correct, but EsiTypes is defined twice in the enum which is of course not possible.

As expected, the related implementation for the other table type is also duplicated:

```rust
impl Related<super::esi_types::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::EsiTypes.def()
    }
}

impl Related<super::esi_types::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::EsiTypes.def()
    }
}
```

I dug around in the API docs a bit, but I can't seem to find a way to make this schema work. Is there a mechanism to allow this in sea_orm?

**Comment by tyt2y3** [comment]
Hi there, welcome!

The codegen is a little dumb now indeed.

Basically, it can't be done with `Related`, but only the `Linked` trait. The difference is, a pair of Entity can have multiple links, but only one relation.

https://github.com/SeaQL/sea-orm/blob/a6dfb417e2bbb151b8f3432139dc3d1679d70b18/src/tests_cfg/entity_linked.rs#L4-L17

https://github.com/SeaQL/sea-orm/blob/a6dfb417e2bbb151b8f3432139dc3d1679d70b18/src/query/join.rs#L286-L301

See also #89.

Oh btw, is that a Battleship game?

**Comment by tyt2y3** [comment]
Did that help?

**Comment by graysonhead** [comment]
Sorry for the delayed reply, been a bit busy lately to work on side projects. It looks like exactly what I'm looking for, and I'll give it a shot once I get a chance to work on this (hopefully in the next few days.)

Its a PVP statistics tracking system for the MMO Eve online.

Thanks for the help!

**Comment by penso** [comment]
@tyt2y3 I think you should make it bold and very clear somewhere in the documentation: `The difference is, a pair of Entity can have multiple links, but only one relation.`

**Comment by cryptomeemaw** [comment]
sorry any examples of how this was done, i have a situation where i have 2 columns in one table with a relationship to another, unsure how i define the relationship/link in this case also how would i define which column it is on
